plugins/scenario/hu.bme.mit.gamma.scenario.statechart.generator/src/hu/bme/mit/gamma/scenario/statechart/generator/StatechartGenerator.xtend

```diff
@@ -0,0 +1,952 @@
+		}
+
+		statechart.transitions.add(forwardTransition)
+		statechart.transitions.add(violationTransition)
+		statechart.transitions.add(cycleTransition)
+
+		if (nonDeclaredMessageMode == 1) {
+			var binary = createBinaryTrigger
+			binary.leftOperand = forwardTransition.trigger
+			binary.rightOperand = getBinaryTriggerFromTriggers(createOtherNegatedTriggers(set), BinaryType.AND)
+			binary.type = BinaryType.AND
+			forwardTransition.trigger = binary
+		}
+
+		violationTransition.sourceState = tmpChoice
+		if (mod.equals(ModalityType.COLD)) {
+			violationTransition.targetState = coldViolation
+		} else {
+			violationTransition.targetState = hotViolation
+		}
+
+		handleArguments(set.modalInteractions, forwardTransition);
+		if (singleNegetedSignalWithArguments) {
+			var signal = (set.modalInteractions.get(0) as NegatedModalInteraction).modalinteraction as Signal
+			if (!signal.arguments.empty) {
+				var tmp = violationTransition.targetState
+				violationTransition.targetState = forwardTransition.targetState
+				forwardTransition.targetState = tmp
+				forwardTransition.trigger = negateEventTrigger(forwardTransition.trigger)
+
+			}
+		}
+		statechart.regions.get(0).stateNodes.add(state)
+		previousState = state
+
+		return
+	}
+
+	def setupForwardTransition(ModalInteractionSet set, InteractionDefinition first, boolean reversed,
+		boolean isNegated, Transition forwardTransition) {
+		var Trigger t = null
+		if (set.modalInteractions.size > 1) {
+			t = getBinaryTrigger(set.modalInteractions, BinaryType.AND, reversed)
+		} else {
+			t = getEventTrigger(first, reversed)
+		}
+		if (!isNegated) {
+			for (mi : set.modalInteractions) {
+				var a = getRaiseEventAction(mi, !reversed)
+				if (a !== null)
+					forwardTransition.effects.add(a)
+			}
+			forwardTransition.trigger = t
+		} else {
+			forwardTransition.trigger = negateEventTrigger(t)
+		}
+	}
+
+	def handleDelays(ModalInteractionSet set) {
+		var delays = set.modalInteractions.filter[it instanceof Delay]
+		if (!delays.empty) {
+			var delay = delays.get(0) as Delay
+			var td = createTimeoutDeclaration
+			td.name = "delay" + timeoutCount++
+			statechart.timeoutDeclarations += td
+			var ts = createTimeSpecification
+			ts.unit = TimeUnit.MILLISECOND
+			ts.value = delay.minimum.clone
+			var a = createSetTimeoutAction
+			a.timeoutDeclaration = td
+			a.time = ts
+			if (previousState instanceof State)
+				previousState.entryActions += a
+		}
+	}
+
+	def handleArguments(EList<InteractionDefinition> set, Transition t1) {
+		var signals = set.filter[it instanceof Signal].filter[!(it as Signal).arguments.empty]
+
+		if (signals.empty) {
+			if (set.size == 1 && set.get(0) instanceof NegatedModalInteraction &&
+				(set.get(0) as NegatedModalInteraction).modalinteraction instanceof Signal &&
+				!((set.get(0) as NegatedModalInteraction).modalinteraction as Signal).arguments.empty) {
+				signals = set.filter[set.size == 1].filter[it instanceof NegatedModalInteraction].filter [
```

Comment:
Use `filter(NegatedModalInteraction)` instead of `filter[it instanceof NegatedModalInteraction]`.